### DISCUSSION TITLE
[Bug] It is not a bug. It is a help request

### DISCUSSION THREAD

**Discussion by alappatsj** [0nelight/macOS-Disable-RAM-Areas#7] [Q&A] [answered]
I have a Macbook Pro 2015 A1502 13 inch 8 GB RAM & 128 GB SSD model. From last one month, I am facing frequent auto-restart issues & on Apple Diagnostics (D), it was showing PPM002 possible memory module issue. Also gives an error message each time the device turned on. 100% chance for Memory issues. 

sometime the device works perfect for one hour or two. But sometimes it instantly gets restarted. Cannot install any OS, it won't complete. 

I have gone through your Page, it is a very good initiative & if you could help in this kind of case, a lot of poor people can save money by using ram module issue devices again, instead of throwing it and buying new. 

I don't have much knowledge in the coding or these things, but still I can manage with a guide or videos. If you can manage a step by step guide or video (from memtest to refind boot to efi script ). It will help lot of people like me. The problem you are solving is big, a lot of people might be facing this issue.

From whatever I understood, I tried to do the memtest, I don't know i did correctly or not. here is the image. I would really appreciate if you could plz help me on this.

Thank You very much in advance.

![20240915_034209](https://github.com/user-attachments/assets/ade4a305-b9f2-4b94-b2c4-5e2ae7546483)

**Comment by 0nelight** [comment]
Yes! This looks like the ram error which this program might fix.

Please follow [this Guide](https://github.com/0nelight/macOS-Disable-RAM-Areas/issues/1#issuecomment-2339234540) to switch into the error-summary mode within memtest86+ and post screenshot again :) Few right presses on the Keyboard should do it!

**Comment by alappatsj** [comment]
Thank you very, very much for your instant & prompt response. And I do really appreciate the efforts you have taken to help people like us. 

I tried Passmark memtest86. I tested 3 times & all the 3 times, the tests failed inbetween saying "Test has aborted due to too many errors". 

As I mentioned, I have made 3 attempts. In all 3 attempts, the highest error address is same. But out of 3, one time the lowest error address is different (slightly). So when you help me to get the EFI file, please use the lowest number in those 2 numbers. 

Here I have attached the screenshots of all 3 tests. I would really appreciate if you could please provide me the edited output (EFI) required for this particular system. 

I am trying to learn the rEFInd thing. Meantime, could you please help me your file?


![20240915_141319](https://github.com/user-attachments/assets/83f1f842-5c4d-4a37-b8e0-3fbbd93323fe)

![20240915_140439](https://github.com/user-attachments/assets/8e324a21-01f4-4763-a017-017f22dbd2e9)
![20240915_142119](https://github.com/user-attachments/assets/e9b3d65b-a799-4183-9b0f-4ca02251f318)

Lower address seems to be different here

![20240915_141334](https://github.com/user-attachments/assets/def42315-a5a3-4314-8d9c-1389fe6b6449)

**Comment by 0nelight** [comment]
So you certainly need to do some reading and learn yourself a little bit about how to boot into rEFInd, put disable-ram-area.efi into the rEFInd Boot Partition, put the disable-ram Boot Stanza within refind.conf and then adjust the parameters to your lowest and highest defective RAM area.

You should please read the "Steps" in the [Readme](https://github.com/0nelight/macOS-Disable-RAM-Areas/blob/main/README.md) of this Repository - all Links there lead to more Information about how to do the above steps.

I also will transform this Issue into a Q&A Discussion.

The disable-ram-area.efi File you can download on [Release-Page](https://github.com/0nelight/macOS-Disable-RAM-Areas/releases/tag/v0.2.1-beta).

Don't hesitate to ask Questions which are not answered by the above resoures :)

What seems a little bit strange is that memtest86+ (not passmark) reports from 5.8 GB onwards while memtest86 (from passmark) shows from somewhere 6GB onwards. Maybe run memtest86+ (not the one from passmark) again in the Error-Summary Mode. I hope the range of defective RAM is not chaning too much between runs of memtest86+.

**Reply by alappatsj** [reply]
Here I have attached the memtest86+ result.

![20240915_155440](https://github.com/user-attachments/assets/b32cafba-e657-4322-aae2-e16f2c036e55)


About your last message.

Dear, I am a marketing person, I mean, I am not from the IT industry. I have very very limited knowledge about linux or boot or other things mentioned in your guide. Not just me, 1000s of people who may have this issue might not be a tech guy. Your existing guide can help only tech guys. But still to revive this dead device, I would like to try whatever I can. 


 -  boot into rEFInd: 
I managed to install rEFInd with a youtube video guide & I am now able to boot into rEFInd. https://www.youtube.com/watch?v=C5G9tr1JS_Y



- Put/Copy "disable-ram-area.efi" into the rEFInd Boot Partition: I don't know how to do it. I cannot see any guide or video for that. I would really appreciate if you could share any link (for video guide) to see how to add.

- Put/Copy "disable-ram Boot Stanza" within refind.conf:

 Unable to understand what is Stanza. And where to find refind conf and how to do that.

- and then adjust the parameters to your lowest and highest defective RAM area.

Your release is an EFI file. It cannot be opened in textedit. So you have to help non-tech guys like us to know how to edit EFI files, the software required for it etc. As I said, you have made a very good tool but it can only be used by very experienced software developers. 

Whatever you have written in your page is targeting only & only the people who can build a tool like this, not the people who wants to use a tool to revive their device. 

IF possible, help us to fix this. So others can also use this thread to fix their system with your tool.

**Reply by 0nelight** [reply]
Thanks for reaching out!

Okay, I understand you have no background in IT. I will think about how to setup a guide for non-IT people :) your questions already help me doing this.

I will keep you updated here with additional information as soon as I have some time.

**Reply by 0nelight** [reply]
In the meantime:
In your [video](https://www.youtube.com/watch?v=C5G9tr1JS_Y) you posted at 08:46 - you see that the guy has mounted the Bootloader-Partition (the EFI Partition) - exactly what you also have to do.

What you want to do is to copy the disable-ram-areas.efi into this directory and then add a Boot Stanza (a Boot Entry) at the end of textfile of refind.conf.

The Boot Entry to add (aka Boot Stanza) into the textfile "refind.conf" looks like this:
```
menuentry "Disable RAM-Areas with args" {
    icon /EFI/BOOT/icons/vol_efi.png
    loader /EFI/tools/disable-ram-area.efi
    options "0x300000000 0x309FFFF48 3000"
}
```

You used windows to create an refind usb drive? Or how did you create it?

**Reply by 0nelight** [reply]
Also please keep in mind that this tool is only suitable when your defective ram area is reasonably low in size. It is strange that memtest86+ (not from passmark) and memtest86 (from passmark) report different areas.

**Reply by 0nelight** [reply]
**Where to find refind.conf?**
See your [yt video](https://youtu.be/C5G9tr1JS_Y?si=PD-dZEJRDULqNBPC) at 09:31:

![image](https://github.com/user-attachments/assets/bb51b8d8-c733-45ce-a321-7a087edce77b)

**Reply by 0nelight** [reply]
In case you want to do this on windows - make sure you have watched and understand this video of how to mount the EFI Partition on Windows:
[How to mount EFI Partition on Windows](https://youtu.be/y7UWOxsmQy0?si=J-FwoaOomsRWBtks)

**Reply by alappatsj** [reply]
Thank you very much.

With YOUR HELP  & that video, now I have rEFInd boot & also I am able to access the EFI folder.

I created the rEFInd USB drive in Macbook. Just copy the folder into it and done whatever is mentioned in that video. 

1) In the Macbook finder, I can see EFI folder when I mount EFI through Terminal. But in the root, I can see Bootlog file/img & one EFI folder. And inside that main EFI folder, there are multiple folders like Apple, Boot, Microsoft (bootcamp), refind, and tools. 

So My question/doubt is, I have to copy your file (disable-ram-areas.efi) to Main root folder or inside the main root folder where there are many folders including refind folder and other OS folders OR I have to paste into the "refind folder"? 

I assume (correct me if I am wrong), I have to copy your release EFI file to inside refind folder where icons, keys & vars folders available and also redind_x64.efi and refind.conf files are present, right? In simple words, the question is, we have to copy your release file (EFI) to which folder?

2) I have to edit/add 4 lines in refind.conf, right? in the bottom, right? The text you have mentioned in your previous answer is same for all or it is specific to my case or we have to generate it? There are some paths (URL), and also the values given in the last line "options". Is this OPTIONS are the 3 paramters you have mentioned in your guide ?( Lowest Error Address (as reported by memtest86+), Highest Error Address (as reported by memtest86+),and Wait time in milliseconds before the program exits).

3) Do we have to make any changes in the EFI file (your release file)? We have to make changes ONLY & ONLY in rEFInd conf? In simple words, I have to just copy your release EFI file to rEFInd folder and then edit the rEFInd.conf file and add just 4 lines of code ? 

I think if I succeed to fix this, other novice people like can use this thread as guide to fix their system with your tool. I really appreciate & thank you for your efforts & supports.

**Reply by 0nelight** [reply]
> I assume (correct me if I am wrong), I have to copy your release EFI file to inside refind folder where icons, keys & vars folders available and also redind_x64.efi and refind.conf files are present, right?

Yes, I think that is a good place to get the path right.

> I have to edit/add 4 lines in refind.conf, right? in the bottom, right? 

Yes - add! yes, at the bottom.

> There are some paths

Altough not an URL per-se, yes but these are (hierachical) paths. This is the way you tell refind where to get the data from. For an icon for example, you have to specify here a path to an icon. Or for the disable-ram-area.efi its path. so instead of ../refind/icons/icon.ico it will be (in case you copied it under ../refind/) it will be ../refind/disable-ram-area.efi.

> Is this OPTIONS are the 3 paramters you have mentioned in your guide ?

Yes

> Do we have to make any changes in the EFI file (your release file)? 

No, you shouldn't make any changes there. Redownload it if you have, because it will malfunction on the slightest change you make in there.

> In simple words, I have to just copy your release EFI file to rEFInd folder and then edit the rEFInd.conf file and add just 4 lines of code ?

Yes. But you have to adjust the path to the disable-ram-area.efi and of course your memory addresses.

```
menuentry "Disable RAM-Areas with args" {
    icon /EFI/BOOT/icons/vol_efi.png   <--- (1)
    loader /EFI/BOOT/disable-ram-area.efi <--- (2)
    options "0x300000000 0x309FFFF48 3000" <--- (3)
}
```

In case you get Icons correctly displayed in refind, you can use this path (1) from another Boot Entry in the refind.conf to counstruct your loader-path (2) without the "icons" - in case you copied disable-ram-area.efi to "refind". May be possible that "BOOT" is called "refind" on your side.

Under options (3) you have to adjust the values according to the lowest and highest error output you get in memtest86+. 

Please keep me updated about success or non-success - for a Marketing Guy you do good Job regarding low-level System configuration ^^

In case something doesn't work please provide - as before - as much information as possilbe. 
Especially getting the loader-path right is important. 

In case you copy the above Boot Stanza; remove all "<--- (..)" .. of course before putting into refind.conf; it cannot handle these.

**Reply by alappatsj** [reply]
Thank You very very much for all your support.

I have reached where you want us to reach.  I have successfully reached the page where we get the result of your EFI file (release file). But unfortunately we missed something. But thats ok, we can workout (I mean you have to help me to fix it). But the stage we have reached now is something extra ordinary. Thank You very much.

Here is the photo of the result page.

![20240916_024612](https://github.com/user-attachments/assets/31dca3a3-a880-4cef-a08f-9e3846962438)



This is to see, how I have made the entry or the entries I made in rEFInd conf file.

![20240916_025001](https://github.com/user-attachments/assets/42e4fb82-6899-465c-aaeb-16d0eacd00c6)

![20240916_025044](https://github.com/user-attachments/assets/16859045-53ad-42f0-b75b-77c3d55c2a4b)

**Reply by 0nelight** [reply]
Wow cool! You made it rellay far 😊 Congrats 🥳😎

So on your Side everything looks okay!

Regarding the Error-Message - we have to dig deeper into why this is not working. 

Could you please provide me with the following Information?

For finding your Firmware Version please follow this steps:
1. Click on the Apple logo in the top-left corner of your screen.
2. Select About This Mac.
3. In the window that opens, click on System Report... (this will open the System Information app).
4. In the Hardware Overview section, you’ll find an entry called **Boot ROM Version** or **Firmware Version**.


**System**
Mac-Computer Model - MacBookPro 11'' Late 2099
MacOS System Version - Big Sur 11.1.11
Firmware Version - 111.1.1.1.1
Bootmanager Version - rEFInd 0.99.9
disable-ram-area.efi Release-Version - v0.0.0-alpha

**Output of [Apple Diagnostic](https://support.apple.com/en-us/102550)**
PPM00X - Error Message

**Output of memtest86+ in Error-Summary-Mode**
0x9090010 0x9090020

**Reply by alappatsj** [reply]
System Firmware Version:	489.0.0.0.0
  OS Loader Version:	540.120.3~37
  SMC Version (system):	2.28f7

**Reply by alappatsj** [reply]
Model Name:	MacBook Pro
  Model Identifier:	MacBookPro12,1
  Processor Name:	Dual-Core Intel Core i5
  Processor Speed:	2.7 GHz
  Number of Processors:	1
  Total Number of Cores:	2
  L2 Cache (per Core):	256 KB
  L3 Cache:	3 MB
  Hyper-Threading Technology:	Enabled
  Memory:	8 GB
  System Firmware Version:	489.0.0.0.0
  OS Loader Version:	540.120.3~37
  SMC Version (system):	2.28f7

**Reply by 0nelight** [reply]
This time, please provide only Information as requested; don't put in the rest of device Information (fe. cpu specs) because I can google it and it only makes it harder for other people to find relevant information on this page :)

**Reply by 0nelight** [reply]
I suspect that the Lower Error Address and the highest Error Address reported by your memtest86+ maybe wrong as they are not the same as the one from passmark memtest86 (regarding your previous screenshots).

Could you please give your Passmark memtest86+ Lowest und Highest Addresses a try. Give them as new Options in refind.conf to the disable-ram-area.efi and maybe they are working - please do this first and report about success or non-success here :) 

In case it didn't work with the lowest and highest address reported by the passmark memtest86:

Please try out the following Version of memtest86+ [this version](https://www.memtest.org/download/v7.00/mt86plus_7.00.binaries.zip) - ONLY the **memtest64.efi** file. You have to put it into the TOOLS Folder of refind Efi Partition and it should automatically be visible in the lower row of your refind Bootscreen to lunch.

**Reply by alappatsj** [reply]
Memtest86+ version is v7.0



Passmark memtest86+ V11.0 - Lowest and Highest Addresses:

0x178719250 (6023 MB) 0x17A71C290 (6055 MB)

**Reply by 0nelight** [reply]
Please give the passmark values a try. As before put them into the Options Lines in refind.conf (below the "loader" line for disable-ram-area.efi) and maybe then the utility will work - please do this first and report about success or non-success here :)

**Reply by alappatsj** [reply]
I tried the Passmark Values.
Just changed the lower & higher values with the passmark memtest values. Nothing else. Just changed the values.

What to say bro? I had tried both memtest86+ and also Passmark memtest, and the both utilities used to stop the test in the middle of the tests and announce FAIL. Within short time, they were sure, the ram is very much faulty. 

Do you know what did I get after using your tool? 

Look below.

![20240916_051023](https://github.com/user-attachments/assets/6ee735b4-9730-4b2b-be39-bdff3c8a8432)


After applying it, I boot into the system & opened all the apps in my dock and opened 15 tabs with Youtube 1080p videos. The system still ON, didn't reboot yet. It was not the story prior to Onelight's magic utility. Thank You very much Onelight, it is a great magical tool. 

For those who reading these messages: This utility does work, it does help. Consider me as the witness of this miracle. 

<img width="1280" alt="Screenshot 2024-09-16 at 5 35 57 AM" src="https://github.com/user-attachments/assets/07de5b3e-c625-4fea-91ad-60f5415a1534">

Used RAM reached to 6.3 GB. It used to auto restart before.

Just check the Swap used: Oh God, it is close to 1 GB.

check the dock and also the tabs opened on top of the image.

And this entire message is typed & posted from this very ram faulty laptop.

Thank You very much bro, now we have to automate the fix.

**Reply by alappatsj** [reply]
sorry to post again. I just can't control my excitement. The RAM usage is stable around 6.5 GB, it is not going to 7 GB. But check the Swap used, it is almost 2 GB. The miracle is, the system still working fine, didn't reboot & it is not slow, it is very much smooth as it was booted. No fan noise or anything, perfect device. Thank You bro. 


<img width="1280" alt="Screenshot 2024-09-16 at 5 46 05 AM" src="https://github.com/user-attachments/assets/375c7e4c-e4b7-4100-bfc9-7a956941c7b7">

<img width="1280" alt="Screenshot 2024-09-16 at 6 13 48 AM" src="https://github.com/user-attachments/assets/52ab24b1-f84f-46f9-a6ce-ab8777098ad4">

**Reply by 0nelight** [reply, marked as answer]
Congratulations 😎😊🥳🎉

Glad all the effort you put in was finally worth it 😊😎

It will also help other people reading this Discussion. I will also use the Insights of this Discussion in a future Guide!

Welcome to the club of people who set a tiny amount of their RAM to "unusable" so macOS stays away from this area. 

I see it as a great way to extend hardware lifespan, protect customer and shareholder value, and fight inflation by lowering demand for new devices, thus preserving the purchasing power of the dollar through more efficient resource use. 😊

Keep in mind that you have to run the tool on every boot. Also I have not much data yet of how macOS behaves when going to Hibernation and back again - but I think the ram stays disables.

In case at some point you want to make the utility run automatically every boot, consider reading [this Guide ](https://github.com/0nelight/macOS-Disable-RAM-Areas/discussions/3) of how to do that.

Also it would be nice to [open a "Success-Story"-Issue](https://github.com/0nelight/macOS-Disable-RAM-Areas/issues/new?assignees=&labels=success&projects=&template=success-story.md&title=Success-Story) and Link this Discussion in it 😊.

May I ask where are you from?

**Reply by alappatsj** [reply]
I Hope you are doing well.

The tool does work initially, and then it depends on how much ram is actually corrupt. Both RAM checking softwares showing different values at different times. But your tool doing exactly what you are claiming. What we have to improve is, finding the actual Faulty areas (in RAM). And I also have serious doubt, there can be issues at multiple sectors. Like IN BETWEEN 1 to 2 GB or 7 to 8 GM or in some cases 1 to 2 gb and ALSO 6 to 7 GB. What I am trying to say is, it may not be just a single Low & high, there can be multiple such bad sectors. Your tool currently accepts one range, not multiple range. When you fix one range, the RAM might find another bad sector because the MemTests are closed inbetween if there are lot of errors. It gives one range but this range might not be the final one. Because the test is not done fully. So once we fix one range & then we do the test again, it will give another set of range. IF the RAM has many bad sectors (like mine), we may have to keep doing the test.


About your questions.

"May I ask where are you from?"
India.


"In case at some point you want to make the utility run automatically every boot, consider reading https://github.com/0nelight/macOS-Disable-RAM-Areas/discussions/3of how to do that."

I really want to do it automatically, but your guide is for very, very technical & only expert people can pull it off.

I did understand, we have to download shell.efi file from that link. 

And we also have to create a text file with the save as extension '.nsh'.  And copy the below content to this .nsh file.

echo Running disable-ram-area.efi with Parameters
stall 1000000
fs7:\EFI\tools\disable-ram-area.efi 0x190000000 0x21000000 3000
echo Starting MacOs
stall 1000000
fs5:\5668A9CD-407D-35EB-87EB-59CD393D65B1\System\Library\CoreServices\boot.efi

I do understand the the fs7: line (3rd line) is what the earlier (manual) refind.conf content (LowV, HighV and time) with the path to your release file EFI. But what I don't understand is, what is this fs7: ?  Where to find this path & How? How to find it is fs5 or fs6 or fs7?

and same for the 4th line. Again new path 'fs5'. How to find these things?. It is the path to boot EFI, right? but how to find or get this path? By the way I cannot see any boot.efi file file in entire EFI folder (root folder of main EFI  folder which consists of Apple, Boot, Refind, tools etc folders). I checked each and every folder, but can't find boot.efi file or path.

and to set it as default (default selection token), you gave a link which nobody can understand (except top experts). 

To help more people with your tool, you genuinely have to spend little time to simplify your guide. In the existing guide, you shared the link for 113 page pdf file for shell commands and again one unending page from refind website for making it is default.
 
Yes, you did great with your tool. We do really appreciate it. But you seems to be very lazy with your guide. IF your guide is not understandable, the normal non-tech people has to spend money to hire people who understand your guide. So what is the difference you are making? Nothing. Spend money on software people to understand your guide, instead of paying to hardware guys who can make an ATTEMPT (failure or success) to replace the ram ?

**Reply by 0nelight** [reply]
So fs0, fs1… are the mappings of your filesystems partitions within efi. Depending on your computer, these may be different on your side. 

How to find them? Best is to enter EFI-Shell from rEFInd - you will look like a hacker in the command line - altough you are only a normal person.

This command line allows different commands to put in. In windows this is called cmd.exe, in macOs it is called Terminal - every system has its own set of commands - also EFI has its own commands. 

To find the paths, you have to play around with the commands of the EFI-Shell. If you are not lazy, you can read through the documentation and maybe contribute to this project by writing a guide yourself.

It takes *some* effort to read through the documentations of course.

 Regarding disabling multiple regions - you can simply execute the EFI-App multiple times - every time for a specific RAM-Area you want to mark as disabled. You can simply copy/paste the Line in the NSH script and adjust your RAM-Areas in every Line.  
 
 Regarding your changing RAM-Error-Areas: Depending on how damaged your RAM is, it could off course be that every day, more and more RAM-Areas are becoming defective. But in my personal case, my MacBookPro RAM has only one tiny defective area around the 12G area and it didn’t change last 10 years - neither in size nor in position. In that cases my Tool is a great workaround.

If you can or want afford a Hardware-Repair - of course this is always the better way. This tool is for poor and lazy people like me.   I am really happy to announce my special friend - ChatGPT - which doesn’t need sleep and doesn’t has any other interest than helping you.   

Regarding finding fs1, fs2 and so on within the EFI-Shell I asked him how he could help you. You can find [his answers here](https://chatgpt.com/share/66f26a77-abe0-8010-b317-f4cc75707335) - without you needing to read any documentation :-)

**Reply by alappatsj** [reply]
I do appreciate your efforts to type long answer. Yes, your tool does work exactly what you are claiming "Disabling the RAM areas we input." I somehow managed to make it work manually (does not work always, but does works some/many times (may be RAM is very faulty). But automating the process is too technical & without any simple guide, I do NOT think except you (or people like you) anybody can ever pull it off. Once again, Thanks for your tool, its helpful.

**Reply by alappatsj** [reply]
BTW IF you really wants to write a guide or help people like me who are NOT from IT industry, You may have to simplify the process.

1) We have to create a NEW text file with the below content and save it as ".nsh" file.

echo Running disable-ram-area.efi with Parameters
stall 1000000
fs7:\EFI\tools\disable-ram-area.efi 0x190000000 0x21000000 3000
echo Starting MacOs
stall 1000000
fs5:\5668A9CD-407D-35EB-87EB-59CD393D65B1\System\Library\CoreServices\boot.efi

Keep the 1st & 2nd line  as it is, Do nothing.

The 3rd line is something important the values are dynamic & are different for each device. It consists the Path to the release file (efi file from this repo) and then the lower & higher value received from the Memtest/RamTest.  Keep 3000 as it is. It is just timeout number. So we have to give the path for the disable-ram-efi file, and then the values/range to disable the ram sectors. So how to give the path which EFI shell understand ? Like with fs5 or fs0, fs7 or fs1.
Now you (@0nelight )  or your friend ChatGPT has to help the users here, how to understand or find the path which efi shell understand. 

keep the 4th & 5th line as it is. Do not make any changes.

In the 6th line, the program is asking to BOOT the system. Giving the path to boot file which you can find in refind.conf file. But still you need give the path which efi shell understands, like starting from fs5 o fs7 or fs1.
Now you (@0nelight )  or your friend ChatGPT has to help the users here, how to understand or find the path which efi shell understand. 

But I have one doubt, what if the user/we have have multiple OSs, like windows or any other dual boot or tripple boot OSs ? Instead of asking to boot MacOS, can't we comeback to Refind boot picker? so the user can select the preferred OS.

And to make the script (.nsh file) DEFAULT, you have to (@0nelight ) write something like I wrote this post. The link you shared for making default, can only understand by highly experienced tech guys.

The guide for the manual process is very much simple for new users. They can use this thread to try the manual process. IF you can discuss  & help the automation here, the new users can use this as guide for 2nd step.

**Reply by MeTalOneOEight** [reply]
The script may be simplified by using variables at the top named after the values used.
set -v LOW_MEM 0x12345678
set -v HIGH_MEM 0x23456789

\EFI\tools\disable-ram-area.efi %LOW_MEM% %HIGH_MEM% 3000

Would that be easier for you?
Same with the paths.
Maybe the paths could be automatically determined by some scripting magic, @0nelight may ask ChatGPT for that.

**Reply by alappatsj** [reply]
what do you want to say? IF you are trying to be too over smart, this is not the platform to show it. Or if you are trying to help the users or the  author, you didn't contribute anything. Nobody has problem to understand the Low & High values or that particular line. People like me who do not have any IT knowledge does not understand the SHELL Path like fs5 or fs7. But you simply passed vilified comment.

**Reply by 0nelight** [reply]
Guys, I appreciate your comments but please try to refrain from further comments in case they have the poor quality - like @alappatsj latest ones. 

This is Github where developers come together and not Reddit or something where these kind of poor communication, repetitive expression, expression of demanding entitlement for ANYTHING regading the project or the actions of the creator of the Repo. 

**Feel free to contribute Pull-requests in case you offer code, create a properly formatted Guide or write a great, well thought-through, well formated, well-meaning, polite and high-quality comment here in the discussions.**

In case you are unable to establish the above demands on Quality of your work, please **refrain** from further comments.

#### Regading usablitiy of this tool:

In case people are too lazy or stupid for reading documentation, maybe they shouldn't use this tool. 

They can go an earn some money with cleaning or something instead of spamming the discussions here and wasting my time.

Please refrain from creating multiple, repetitive or demanding Posts or Issues. Otherwise I will start to report you. Thank you.

@MeTalOneOEight 
Thanks for your Idea. What you recommend is using environmental variables. But I don't see how this would make thinks easier. If someone really puts *some* effort into reading and understanding documentation it should be easy to get it working also for non-tech people. 

Regarding automatic scripting - sure I can write a whole seperate bootmanager like rEFInd and everything would go automatically, but I am not interested in doing so.

In case you want to do that, please open a new repo or fork my code. In case it works better than my solution, I am happy to integrate it in my code base.

**Reply by MeTalOneOEight** [reply]
You may not have problems to understand that line, but others maybe.
Beside that it's cleaner scripting, having the user serviceable part at top.

And now to your specific problem:
map is the command to show the device map, there you will see the fs5.
To change to that device you type fs5 as command.
Then to determine the path you use ls or cd.

If you need a more indepth EFI shell tutorial, you may find one suiting your needs with the search engine of your choice.

**Reply by 0nelight** [reply]
@MeTalOneOEight please mention @alappatsj in your post, as this is relevant for him. 

Altough I lost hope because I already told him about "map" in my linked chatgpt conversation - but it seems he is not interested to put the effort into reading and understanding even short information.

**Reply by alappatsj** [reply]
Building a tool is like a building a great bridge to cross a river, but if the builder make it highly inaccessible for all or for most people, the bridge is just a big concrete useless brick. The time spent to reach there is a sheer waste of time. 
 
And for your kind info, IF I don't put interest to read or put effort or waste as many hours, how did I manage to test & give a very detailed review (I have even posted a very lengthy praise) for your script (manual)? Was there any guide to do that? I had to spent hours to search it on youtube & then I made it work. It was not your guide or there was no any guide for non-techy people. 

people like @daniel-nextgti & @VincePuc99 seems to be well experienced tech people, but still they were NOT able to pull this off themselves. And you are asking people to learn Shell by throwing a 113 page documentation to the people who do not even heard about what is shell. And that is just to get a PATH, not to invent any new script or tool. And the most hilarious thing is, sharing a refind webpage as guide to make it default. Not a single non-tech person may have ever installed refind with the guide or texts written in their website. And you are asking the users to use one of their page to understand, how to make it default. LOL. 

Keep your tool with you. I have already wasted enough hours. It is not worth wasting these many hours.  But put a notice on top (it is there now, earlier it wasn't there. But edit it as) "This is for highly experienced tech people who understands EFI or SHELL or people who can WASTE COUNTLESS hours to read whatever unnecessary references/links the author throws at you. "

And Thank You very much for wasting countless hours. 


> Guys, I appreciate your comments but please try to refrain from further comments in case they have the poor quality - like @alappatsj latest ones.
> 
> This is Github where developers come together and not Reddit or something where these kind of poor communication, repetitive expression, expression of demanding entitlement for ANYTHING regading the project or the actions of the creator of the Repo.
> 
> **Feel free to contribute Pull-requests in case you offer code, create a properly formatted Guide or write a great, well thought-through, well formated, well-meaning, polite and high-quality comment here in the discussions.**
> 
> In case you are unable to establish the above demands on Quality of your work, please **refrain** from further comments.
> 
> #### Regading usablitiy of this tool:
> In case people are too lazy or stupid for reading documentation, maybe they shouldn't use this tool.
> 
> They can go an earn some money with cleaning or something instead of spamming the discussions here and wasting my time.
> 
> Please refrain from creating multiple, repetitive or demanding Posts or Issues. Otherwise I will start to report you. Thank you.
> 
> @MeTalOneOEight Thanks for your Idea. What you recommend is using environmental variables. But I don't see how this would make thinks easier. If someone really puts _some_ effort into reading and understanding documentation it should be easy to get it working also for non-tech people.
> 
> Regarding automatic scripting - sure I can write a whole seperate bootmanager like rEFInd and everything would go automatically, but I am not interested in doing so.
> 
> In case you want to do that, please open a new repo or fork my code. In case it works better than my solution, I am happy to integrate it in my code base.

**Reply by daniel-nextgti** [reply]
@alappatsj 

I think you got here too eager to make it work, but not so eager to understand the logic/read, neither the real purpose of creating a git repository.

@0nelight work is for himself. The choice in making this repository public is mainly connected to his personal interests, and not helping anyone else. I'm sure he has other concerns in his life and helping random persons isn't a priority. Is almost as a hobby, or volunteer.

**The last thing you should be doing is criticize him.** 
This is a Work In Progress, and only god knows if it will ever be on a 1-Click solution state.

@0nelight response to you about the fs0 partitions is correct. It`s impossible to know how YOUR computer is configured, thus, giving you a Step By Step on how to make the script work.

I only have to thank all the help that @0nelight has provided me, since it was only a .C file when i got here, and now, it`s not even necessary to compile any file (Yes, at the time, i've searched how to compile the .C file, by myself, using ChatGPT and Gemini - Didn't hurt -  it was fun, took me 2 days).

Your case of BAD RAM might be different from ours, maybe with multiple ranges of corruption. At this stage, the Script only works on one Range. The only real solution for all of us is to replace your RAM/Logic Board.

That's why @0nelight 's work is do important, it is a temporary alternative, but not the definitive solution.

@alappatsj I think a new logic board should cost around ₹ 26.459,00 in India
How much did you payed for @0nelight `s Work?

Dont be rude. Is unfair.

**Reply by daniel-nextgti** [reply]
@alappatsj and FYI

To know how many FS's you have, just record your Macbook screen with a phone when the Shell.efi starts.

It writes all FS's avaiable on your Macbook screen when opening the .efi .


Then, when the shell opens, use commands like CD FS0 or CS FS1 to access the FS
Inside de FS, type dir ou ls to display all the folders

Only this way you'll be able to know that purpose each FS has on YOUR machine.

That's why you need to learn how shell works. No matter if is Linux, MacOS or Windows, they all rely on Shell.
This Github is not a tutorial like the ones on Youtube.

Here's a Baby-Step video on how use commands on Shell. https://www.youtube.com/watch?v=TPF-eIe-ucU

**Reply by 0nelight** [reply]
@daniel-nextgti Thanks for helping out and making things clearer! 😀 

@alappatsj In my view you don't deserve much further help from me. But because my life is about how I am, not about how you or other people are (to quote a [great 🇮🇳 Guru](https://mx.pinterest.com/pin/sadhguru-never-form-an-opinion-on-anyone-how-they-are-in-this-moment-is-all-that-matters--317433473744197240/) of our times), [maybe you can find help on google.](https://letmegooglethat.com/?q=youtube+What+is+EFI+Shell+and+How+to+Use+it) 

You should really consider doing Meditation to get more calm - works wonderfully for me 🧘‍♂️☺️😊. Makes me much more relaxed and at the same time my mind becomes very sharp. Greetings from 🇪🇺 and thanks for your passion towards this project, despite your struggles to express yourself in a nice way. 🙏🏼
